src/ssh/cockpitsshrelay.c

```diff
@@ -946,27 +947,36 @@ do_password_auth (CockpitSshData *data)
   return rc;
 }
 
+/* We don't support unlocking identities within cockpit-ssh so fail here */
 static int
-do_key_auth (CockpitSshData *data)
+prompt_for_identity_password (const char *prompt, char *buf, size_t len,
+                              int echo, int verify, void *userdata)
+{
+  CockpitSshData *data = (CockpitSshData *) userdata;
+  g_autofree gchar *identity = NULL;
+  if (ssh_options_get (data->session, SSH_OPTIONS_IDENTITY, &identity) == SSH_OK)
+    data->problem_error = g_strdup_printf ("locked identity: %s", identity);
+  return -1;
+}
+
+static int
+do_key_identity_auth (CockpitSshData *data)
 {
   int rc;
-  const gchar *msg;
-  const gchar *key_data;
   ssh_key key;
+  g_autofree gchar *identity = NULL;
+  const gchar *msg;
 
-  g_assert (data->initial_auth_data != NULL);
-
-  key_data = cockpit_authorize_type (data->initial_auth_data, NULL);
-  if (!key_data)
+  rc = ssh_options_get (data->session, SSH_OPTIONS_IDENTITY, &identity);
```

Comment:
but seriously, you shouldn't be using `g_autofree` with an ssh_string.  Although the glib allocator has guaranteed `malloc()`-equivalence since a couple months ago (and in practice has worked this way since 2015) it's still not acceptable to assume that `ssh_string_free` is equivalent to `free()`.

Either:
 - `g_strdup()` the result
 - write your own free function and use the `cleanup` attribute directly,
 - use the appropriate glib-provided macros to provide a g_autoptr() wrapping for, ssh_string; or,
 - just don't use autofree here

Reply:
ok .-.